src/app/core/public/api/impl/Component.ts

```diff
@@ -1,45 +1,46 @@
 import {V, Vector} from "Vector";
 
+import {FromConcatenatedEntries} from "core/utils/Functions";
+
 import {Schema} from "core/schema";
 
-import {Component}     from "../Component";
-import {ComponentInfo} from "../ComponentInfo";
-import {Port}          from "../Port";
+import {Component} from "../Component";
+import {Port}      from "../Port";
 
 import {BaseObjectImpl} from "./BaseObject";
-import {PortImpl}       from "./Port";
+import {CircuitState}   from "./CircuitState";
 
 
-export class ComponentImpl extends BaseObjectImpl implements Component {
+export abstract class ComponentImpl<
+    State extends CircuitState = CircuitState
+> extends BaseObjectImpl<State> implements Component {
     public readonly baseKind = "Component";
 
     protected getObj(): Schema.Component {
-        const obj = this.circuit.getCompByID(this.id);
+        const obj = this.internal.getCompByID(this.id);
         if (!obj)
             throw new Error(`API Component: Attempted to get component with ID ${this.id} could not find it!`);
         return obj;
     }
 
-    public get info(): ComponentInfo {
-        throw new Error("Unimplemented");
-    }
+    public abstract get info(): ReturnType<State["constructComponent"]>["info"];
```

Comment:
fancy

Reply:
I hate this tbh, it feels extremely hacky
The alternative I can think of is passing like a bunch of template arguments into Circuit/Component/Wire/etc
like:
`Component<DigitalCircuit, DigitalComponent, DigitalWire, DigitalPort>` to specify each sub-type of Circuit/Component/Wire/Port idk what do you think

Reply:
Maybe not if there isn't another reason to?